### PR TITLE
Rework hierarchy commands

### PR DESCRIPTION
**Pull request by devil-ira**
# Objective

* Improve API consistency (`add_child` vs `add_children` vs `push_children`).
* Add singular variants (eg. `remove_child`).
* Reduce code duplication; The commands were implemented twice, once on `EntityMut` and again with `Command`s. `EntityMut` impls are now re-used.
* A whole load of tests added. `HierarchyEvents` are now also tested.

I renamed `child_builder.rs` to `commands.rs` as it fits better, and because of how unhelpful the git diff got from how much the file changed.

## Changelog
* Merged `BuildChildren` and `BuildWorldChildren` into `HierarchyCommands` and implemented it for both `EntityMut` and `EntityCommands`. 
* Removed `add_children`.
* Renamed `push_children` to `add_children`.
* Removed hierarchy commands from `WorldChildBuilder`. This one was wierd. Check the migration guide. 

## Migration Guide

`add_children` has been removed. If you need to access a value from the closure, it's recommended to avoid using a closure with `with_children` and instead use `add_child` or `set_parent`. 

`push_children` has been renamed to `add_children` to match `add_child`.

`WorldChildBuilder` no longer has hierarchy methods like `push_children`, etc.
```rust
world.spawn().with_children(|parent| {
	// If you were doing this:
	parent.spawn();
	parent.push_children(some_entities);
	// You can do the following instead.
	parent.spawn_empty().add_children(some_entities);
});

### CODE REVIEW

File: crates/bevy_hierarchy/src/commands.rs [hunk 1]
```diff
@@ -0,0 +1,1167 @@
...
+    fn insert_children(&mut self, index: usize, children: &[Entity]) -> &mut Self;
+
+    /// Remove the child.
+    fn remove_child(&mut self, child: Entity) -> &mut Self;
+
+    /// Remove the children.
+    fn remove_children(&mut self, children: &[Entity]) -> &mut Self;
+
+    /// Remove all children.
+    fn remove_all_children(&mut self) -> &mut Self;
```
Comment (james7132) [inline crates/bevy_hierarchy/src/commands.rs:43]: Bikeshed: `clear_children`.

File: crates/bevy_hierarchy/src/components/parent.rs [hunk 2]
```diff
@@ -8,7 +8,8 @@ use bevy_reflect::Reflect;
 use std::ops::Deref;
 
 /// Holds a reference to the parent entity of this entity.
-/// This component should only be present on entities that actually have a parent entity.
+///
+/// Only present on entities that have a parent.
```
Comment (james7132) [inline crates/bevy_hierarchy/src/components/parent.rs:12]: Please make a note on how this interacts with `With` and `Without`: `Without` will find root entities. `With` finds any intermediate or leaf entities.

File: crates/bevy_hierarchy/src/events.rs [hunk 3]
```diff
@@ -3,7 +3,7 @@ use bevy_ecs::prelude::Entity;
 /// An [`Event`] that is fired whenever there is a change in the world's hierarchy.
 ///
 /// [`Event`]: bevy_ecs::event::Event
-#[derive(Debug, Clone)]
+#[derive(Debug, Clone, PartialEq)]
```
Comment (james7132) [inline crates/bevy_hierarchy/src/events.rs:6]: Any reason why this doesn't also implement `Eq`?

File: crates/bevy_hierarchy/src/commands.rs [hunk 4]
```diff
@@ -0,0 +1,1167 @@
...
+
+    /// Insert the children at the given index.
+    ///
+    /// If the child already had a parent it will be removed from that parent.
+    fn insert_children(&mut self, index: usize, children: &[Entity]) -> &mut Self;
+
+    /// Remove the child.
+    fn remove_child(&mut self, child: Entity) -> &mut Self;
+
+    /// Remove the children.
```
Comment (james7132) [inline crates/bevy_hierarchy/src/commands.rs:39]: More complete docs here would be nicer. This seems a bit lacking.

**Review by james7132 — CHANGES_REQUESTED** [review]
This seems to add a bunch of code, but doesn't remove the existing implementation.

**Comment by devil-ira** [comment]
Oops. The old file is no longer imported, but I forgot to remove it after rebasing.

File: crates/bevy_hierarchy/src/commands.rs (hunk 4 above)
Comment (devil-ira) [reply crates/bevy_hierarchy/src/commands.rs:39]: Sure, what would you like to see documented?

**Review by james7132 — APPROVED** [review]
This took a lot longer to review than it should have. LGTM.

**Comment by devil-ira** [comment]
I already split part of this PR into #6727, but I'm going to try to split it further so it's easier to review :)

Apologies to James for making you review this monster 🙏 haha

**Comment by james7132** [comment]
Should we close this and split out it's parts into other PRs?

**Comment by devil-ira** [comment]
Yup!

**Pull request closed without merging**